### ISSUE TITLE
[Open3D WARNING] Read PLY failed: unable to open file: ./datasets/ycb/YCB_Video_Dataset/models_rgb/obj_000002.ply

### ISSUE THREAD

**Issue by ZJU-PLP** [Gorilla-Lab-SCUT/BiCo-Net#1]
Hi, Zelin.

I meet the error when to reproduce your code.  Cloud you mind sharing the **models_rgb** file about YCB dataset?

**Comment by aragakiyui611** [comment]
This folder is modified from `models_fine` of YCBV from bop datasets. You could download these from [bop_datasets/ycbv_models](https://bop.felk.cvut.cz/media/data/bop_datasets/ycbv_models.zip). Thank you for raising this question.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Thanks for your reply.
I follow your step and run the code again. I meet the error as below. Could you mind sharing the version of open3d that you use? I find the **pnc.colors** is wrong.
![Selection_052](https://user-images.githubusercontent.com/32757213/182312478-320c9fa1-c385-493a-a59e-75f64ffb1333.png)

Btw, the linemod dataset also has similar problem:
![Selection_053](https://user-images.githubusercontent.com/32757213/182312842-ba7a26c0-3a7d-4ca4-9c97-2044a0cde577.png)

**Comment by ZJU-PLP** [comment]
@aragakiyui611 I fixed the problem by setting pnc = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), **format='xyzn'**) and pnc_cp = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), **format='xyzrgb'**) to get xyz, normal and color information of cloud.

**Comment by aragakiyui611** [comment]
My open3d==0.9.0.0

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Okkk, thanks for your reply.

When I run the code for linemod dataset, I meet the error:
![image](https://user-images.githubusercontent.com/32757213/182626198-5936a6d4-071b-4385-ba35-56a8d147e48e.png)

Could you mind giving some advice?

**Comment by aragakiyui611** [comment]
```
            if len(data) == 1: 
                pts_dis.append(np.inf)       
                result_obj_id.append(data[0])
                continue
```
please add thsese lines before line 99 in the train.py or other flexible solution.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 I follow your advice and run the code again. However, I still meet the error.
![image](https://user-images.githubusercontent.com/32757213/182774353-fb3929e0-8e8b-4486-a7d7-c02f343b7a04.png)

**Comment by ZJU-PLP** [comment]
@aragakiyui611 I find that opt.batch_size cannot adjust the dataset loading for improving train speed. Maybe ([Line 65](https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/main/train.py#L65)) batch_size =1 should changed to be batch_size=opt.batch_size ? If not, the actual batch_size is always to be 1 and gpu utilizing efficiency is very slow.

**Comment by aragakiyui611** [comment]
> @aragakiyui611 I find that opt.batch_size cannot adjust the dataset loading for improving train speed. Maybe ([Line 65](https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/main/train.py#L65)) batch_size =1 should changed to be batch_size=opt.batch_size ? If not, the actual batch_size is always to be 1 and gpu utilizing efficiency is very slow.

This code is modified from Densefusion which uses the same training strategy. The `opt.batch_size` is the final batch size when backward. The reason of using forward batch size=1 is that the input cropped images size are different, which may affect the efficiency a bit. You could try resize/pad all cropped image to the same size and stack them to a batch but accuracy may drops.

> @aragakiyui611 I follow your advice and run the code again. However, I still meet the error. ![image](https://user-images.githubusercontent.com/32757213/182774353-fb3929e0-8e8b-4486-a7d7-c02f343b7a04.png)

As far as this, I suggest check the len of returned tuple and if the size not meet the unpack size, then just continue the next loop, solutions are flexible. There are different solutions and I just provide the  idea. Thank you!

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Ok, thanks four your patient reply again. I will have a try.

Another several problems:

1. I want to know how long to train epoch=50 in terms of three different datasets(YCB, LineMod, Occlusion LineMOD) and the GPU device?

2. I cannot visualize the the training curve by using the dash command. I want to know whether you have met this problem?

```
cd runs  
tensorboard --logdir=./
```

3. Could you mind sharing the experiment code details about Occlusion LineMOD? I cannot find the training code and eval code in your published version.

**Comment by aragakiyui611** [comment]
1. You can see the log and calculate the training time, for LineMOD dataset it cost around 4 hours to train 30 epochs. For YCBV 24h to train 50 epoch.
2. I can lauch my tensorboard with `tensorboard --logdir=./` , this might be you env problem.
<img width="904" alt="image" src="https://user-images.githubusercontent.com/74336517/183234237-5f27a2f8-5655-4dc1-b2e8-6d2ab770835d.png">

**Comment by ZJU-PLP** [comment]
@aragakiyui611 
1. For **YCBV 24h to train only 5 epoch** in the Nvidia 3090Ti (in our lab server) and For **YCBV 24h to train only 4 epoch** in the Nvidia 1080Ti in my local desktop computer. Could you mind sharing **the type of Nvidia GPU device**  for training? The training log is as below:

![image](https://user-images.githubusercontent.com/32757213/183235457-1fc27e44-030b-4e6f-963a-c787f18c6225.png)
![image](https://user-images.githubusercontent.com/32757213/183235459-fdc7fe1e-c325-442e-8438-da6adbba8e60.png)


2. I update the tensorboard version and the page is to be right.

**Comment by yunnan66** [comment]
> @aragakiyui611 Okkk, thanks for your reply.
> 
> When I run the code for linemod dataset, I meet the error: ![image](https://user-images.githubusercontent.com/32757213/182626198-5936a6d4-071b-4385-ba35-56a8d147e48e.png)
> 
> Could you mind giving some advice?

@ZJU-PLP Hello.
I have the same problem as you. Have you solved this problem? How was it solved?

**Comment by ZJU-PLP** [comment]
@yunnan66

 Yes, I have solved this problem by adding  the code in Line 99 and Line 143 as below in train and eval step of train.py :

 ```
 if len(data) == 1:
      continue
```

 ```
 if len(data) == 1:
      pts_dis.append(np.inf)
      result_obj_id.append(data[0])
      continue
```

**Comment by yunnan66** [comment]
> @yunnan66
> 
> Yes, I have solved this problem by adding the code in Line 99 and Line 143 as below in train and eval step of train.py :
> 
> ```
> if len(data) == 1:
>      continue
> ```
> 
> ```
> if len(data) == 1:
>      pts_dis.append(np.inf)
>      result_obj_id.append(data[0])
>      continue
> ```

@ZJU-PLP Thank you for your reply. I will try.

**Comment by yunnan66** [comment]
@ZJU-PLP Hello.
May I ask how you solved the following problem?
![image](https://user-images.githubusercontent.com/90179991/187034969-ac6fc56c-9bc6-440b-b9d4-223260dd114a.png)

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Hi, I solved this problem by replacing dataset.py (ycb and linemod, [Line62-Line66](https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/main/datasets/ycb/dataset.py#L62-L66)) with code of below:

```
pnc = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), format='xyzn')
pnc_cp = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), format='xyzrgb')
pp = np.array(pnc.points)/1000.0# (N, 3)
pp = pp - pp.mean(0)
nn = np.array(pnc.normals)# (N, 3)
cc = np.array(pnc_cp.colors)# (N, 3) normalized
```

**Comment by yunnan66** [comment]
> @aragakiyui611 Hi, I solved this problem by replacing dataset.py (ycb and linemod, [Line62-Line66](https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/main/datasets/ycb/dataset.py#L62-L66)) with code of below:
> 
> ```
> pnc = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), format='xyzn')
> pnc_cp = o3d.io.read_point_cloud('%s/models_rgb/obj_%06d.ply'%(self.root, cad_id), format='xyzrgb')
> pp = np.array(pnc.points)/1000.0# (N, 3)
> pp = pp - pp.mean(0)
> nn = np.array(pnc.normals)# (N, 3)
> cc = np.array(pnc_cp.colors)# (N, 3) normalized
> ```

@ZJU-PLP Thank you very much for your reply. I have successfully solved this problem with your method.

Another several problems:

1.I don't know if you have encountered this problem, I simply set num_works to 0, resulting in slow training speed, do you have a better solution?
![QQ图片20220829094809](https://user-images.githubusercontent.com/90179991/187107850-648583a1-a44e-4cee-9ffd-280cae474eae.jpg)

2.For LM 30h to train 50 epoch in the Nvidia 3090Ti, the training speed is slow, probably because the crop size of each picture is inconsistent and batch_size is set to 1. Do you have a better way to deal with it?

3.Do you have an implementation for the LM-O dataset? How does it work?

**Comment by ZJU-PLP** [comment]
@aragakiyui611 

1. The num_works should not be set 0. You can look up  the function of torch.utils.data.DataLoader about num_workers.
2. I also want to solve this problem about the speed of training. However, I have not found the nice method. 
3. No, I have not done this work. Furthermore, I have proposed the issue in #2 . However, the author have not answered this issue up to now.

**Comment by TerenceHsu666** [comment]
> > @yunnan66
> > Yes, I have solved this problem by adding the code in Line 99 and Line 143 as below in train and eval step of train.py :
> > ```
> > if len(data) == 1:
> >      continue
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > if len(data) == 1:
> >      pts_dis.append(np.inf)
> >      result_obj_id.append(data[0])
> >      continue
> > ```
> 
> @ZJU-PLP Thank you for your reply. I will try.

We fixed this bug in the training script.

**Comment by ZJU-PLP** [comment]
@CavendyHsu Thanks for your updating.

Furthermore, could you mind sharing the script if it is possible for plotting Figure 3 in terms of different levels of occlusion on the YCB-Video benchmark? I have researched many papers but could not get the script code.

**Comment by Karthus0930** [comment]
@ZJU-PLP Hello，may I ask you some questions about the speed of training and how to Implement visual code on Linemod dataset

**Comment by ZJU-PLP** [comment]
> @ZJU-PLP Hello，may I ask you some questions about the speed of training and how to Implement visual code on Linemod dataset

1. I spent about 39h/39 epoches by GTX 2080Ti in LMO dataset and 51h/50 epoches by GTX 2080Ti in LM dataset.
2. I had not implemented visual code on Linemod dataset. Then, I will try to implement this part from PVD3D visual code.

**Comment by Karthus0930** [comment]
May I have your WeChat that we can share each other?



发自我的iPhone


------------------ Original ------------------
From: Linpeng Peng ***@***.***&gt;
Date: Tue,Nov 8,2022 9:27 PM
To: Gorilla-Lab-SCUT/BiCo-Net ***@***.***&gt;
Cc: Karthus0930 ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [Gorilla-Lab-SCUT/BiCo-Net] [Open3D WARNING] Read PLY failed:unable to open file:./datasets/ycb/YCB_Video_Dataset/models_rgb/obj_000002.ply (Issue #1)

**Comment by ZJU-PLP** [comment]
> May I have your WeChat that we can share each other? 发自我的iPhone
> […](#)
> ------------------ Original ------------------ From: Linpeng Peng ***@***.***&gt; Date: Tue,Nov 8,2022 9:27 PM To: Gorilla-Lab-SCUT/BiCo-Net ***@***.***&gt; Cc: Karthus0930 ***@***.***&gt;, Comment ***@***.***&gt; Subject: Re: [Gorilla-Lab-SCUT/BiCo-Net] [Open3D WARNING] Read PLY failed:unable to open file:./datasets/ycb/YCB_Video_Dataset/models_rgb/obj_000002.ply (Issue #1)

Ok, so your WeChat ID is?
